data_rentgen/server/providers/auth/keycloak_provider.py

```diff
@@ -103,25 +106,42 @@ async def get_current_user(self, access_token: str, *args, **kwargs) -> User:
             raise AuthorizationError("Invalid token payload")
         return await self._uow.user.get_or_create(UserDTO(name=login))  # type: ignore[arg-type]
 
+    async def logout(self, refresh_token: str):
+        try:
+            return self.keycloak_openid.logout(refresh_token)
+        except KeycloakConnectionError as err:
+            logger.error("Error when trying to get token: %s", err)
+            return None
+
     def decode_token(self, access_token: str) -> dict[str, Any] | None:
         try:
             return self.keycloak_openid.decode_token(token=access_token)
         except Exception as err:
             logger.info("Access token is invalid or expired: %s", err)
             return None
 
-    def refresh_access_token(self, refresh_token: str) -> tuple[str, str]:  # type: ignore[return]
+    def refresh_access_token(self, refresh_token: str, origin_url: str) -> tuple[str, str]:  # type: ignore[return]
         try:
             new_tokens = self.keycloak_openid.refresh_token(refresh_token)
             logger.debug("Access token refreshed")
             return new_tokens.get("access_token"), new_tokens.get("refresh_token")
         except Exception as err:
             logger.debug("Failed to refresh access token: %s", err)
-            self.redirect_to_auth()
+            self.redirect_to_auth(origin_url)
 
-    def redirect_to_auth(self) -> None:
-        auth_url = self.keycloak_openid.auth_url(
-            redirect_uri=self.settings.keycloak.redirect_uri,
-            scope=self.settings.keycloak.scope,
-        )
-        raise RedirectError(message=auth_url, details="Authorize on provided url")
+    def redirect_to_auth(self, state: str = ""):
+        try:
+            state = base64.b64encode(state.encode("utf-8"))  # type: ignore[assignment]
+
+            auth_url = self.keycloak_openid.auth_url(
+                redirect_uri=self.settings.keycloak.redirect_uri,
+                scope=self.settings.keycloak.scope,
+                state=state.decode("utf-8"),  # type: ignore[attr-defined]
+            )
+
+        except KeycloakConnectionError as err:
+            logger.error("Failed connect to Keycloak: %s", err)
+            # TODO: What exception should we raise here?
+            raise HTTPException(status_code=HTTPStatus.BAD_REQUEST, detail=err)
```

Comment:
Do we really need to catch this exception? We already have exception handler for all unknown exceptions raised by application which returns 500 error

Reply:
500 is bad ((

Ok i remove this. For me 500 was excpetion in our code. But somethimes keycloak server can be unreacheable

Reply:
If we cannot connect to Keycloak (e.g. it is unreachable), what should we do instead of 500?